DeviceSDK/src/main/java/org/astarteplatform/devicesdk/protocol/AstarteInterfaceMapping.java

```diff
@@ -67,25 +30,31 @@ public Type getType() {
     return type;
   }
 
-  boolean isTypeCompatible(Type otherType) {
-    // Here is our long and boring compatibility map.
-    for (Collection<Type> compatibilityList : mTypeCompatibilityList) {
-      if (compatibilityList.contains(otherType) && compatibilityList.contains(getType())) {
-        return true;
-      }
+  protected boolean isTypeCompatible(Type otherType) {
+    return (getType() == otherType);
+  }
+
+  public void validatePayload(Object payload) throws AstarteInvalidValueException {
+    if (!isTypeCompatible(payload.getClass())) {
+      throw new AstarteInvalidValueException(
+          String.format(
+              "Value incompatible with parameter type for %s: %s expected, %s found",
+              getPath(), getType(), payload.getClass()));
+    }
+    if (payload instanceof Double && ((Double) payload).isNaN()) {
+      throw new AstarteInvalidValueException(
+          String.format("Value per %s cannot be NaN", getPath()));
     }
-    return false;
   }
 
-  boolean isTypeCompatible(String astarteType) {
-    // Here is our long and boring compatibility map.
-    for (Collection<Type> compatibilityList : mTypeCompatibilityList) {
-      if (compatibilityList.contains(stringToJavaType(astarteType))
-          && compatibilityList.contains(getType())) {
-        return true;
-      }
+  public void validatePayload(Object payload, DateTime timestamp)
+      throws AstarteInvalidValueException {
+    validatePayload(payload);
+    if (timestamp != null) {
+      // FIXME: Maybe go easier, and just throw a warning?
```

Comment:
s/FIXME/TODO/.

Reply:
I totally agree. It was an old comment I found pre refactoring. Let me know the wanted behaviour and I'll fix it